### PR TITLE
tor: respect non-onion -onlynet= for outgoing Tor connections

### LINKED ISSUE #22647: If both options "onion" and "proxy" are unset, no outbound Onion connections should be made
bitcoind 22.0rc2

Having the following `bitcoin.conf`
```
upnp=0
debug=i2p
shrinkdebugfile=1
onlynet=i2p
listen=1
timeout=20000
server=1
maxtxfee=0.001
i2psam=127.0.0.1:7656
addnode=72l3ucjkuscrbiiepoehuwqgknyzgo7zuix5ty4puwrkyhtmnsga.b32.i2p:0
addnode=c4gfnttsuwqomiygupdqqqyy5y5emnk5c73hrfvatri67prd7vyq.b32.i2p:0
addnode=gehtac45oaghz54ypyopim64mql7oad2bqclla74l6tfeolzmodq.b32.i2p:0
addnode=h3r6bkn46qxftwja53pxiykntegfyfjqtnzbm6iv6r5mungmqgmq.b32.i2p:0
addnode=hnbbyjpxx54623l555sta7pocy3se4sdgmuebi5k6reesz5rjp6q.b32.i2p:0
addnode=pjs7or2ctvteeo5tu4bwyrtydeuhqhvdprtujn4daxr75jpebjxa.b32.i2p:0
addnode=wwbw7nqr3ahkqv62cuqfwgtneekvvpnuc4i4f6yo7tpoqjswvcwa.b32.i2p:0
addnode=zsxwyo6qcn3chqzwxnseusqgsnuw3maqnztkiypyfxtya4snkoka.b32.i2p:0
```
Makes bitcoind connect outgoing to Onion network.
`bitcoind --help` says for option `onlynet`:
_"if it is used with non-onion networks and the -onion or -proxy option is set, then outbound onion connections will still be made; use -noonion or -onion=0 to disable outbound onion connections in this case."_

But as you can see, neither `onion` nor `proxy` is set. `proxy` defaults to "disabled" (=0 ?), and `onion` defaults to value of `proxy`, so if `proxy` is unset=disabled, `onion` should be disabled(=0), too. There should be no need for setting `onion=0` in this case.

The above sentence would be clearer like so: _"use -noonion or -onion=0 to disable outbound onion connections if proxy is set"_

**Expected behavior**

No outgoing connections to Onion network, only to I2P network.

**Actual behavior**

Outgoing connections to both Onion and I2P network.

**To reproduce**

Use above bitcoin.conf.

**System information**

Debian Buster amd_64

### PR DESCRIPTION
**Pull request by vasild**
If
* `-onlynet=` is given one or more times but none of them includes
  `onion` and
* `-proxy` is not set and
* `-onion` is not set,

then we should not be making outbound connections to the Tor network.

Fixes https://github.com/bitcoin/bitcoin/issues/22647

### CODE REVIEW

**Comment by vasild** [comment]
See also https://en.wikipedia.org/wiki/Principle_of_least_astonishment

**Comment by jonatack** [comment]
Concept ACK. It may be good to make it a separate function with test coverage.

File: src/torcontrol.cpp [hunk 1]
```diff
@@ -373,9 +373,18 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
...
+        if (gArgs.IsArgSet("-onlynet")) {
+            allow_outbound_onion = false;
+            for (const auto& net : gArgs.GetArgs("-onlynet")) {
+                if (ParseNetwork(net) == NET_ONION) {
+                    allow_outbound_onion = true;
+                    break;
+                }
+            }
+        }
+        if (gArgs.GetArg("-onion", "") == "" && allow_outbound_onion) {
```
Comment (jonatack) [inline src/torcontrol.cpp:387]: A function seems nicer than a localvar here (optionally declare in header file for unit testing)
```diff
+bool AllowOutboundOnion()
+{
+    if (gArgs.IsArgSet("-onlynet")) return false;
+    for (const auto& net : gArgs.GetArgs("-onlynet")) {
+        if (ParseNetwork(net) == NET_ONION) return true;
+    return false;
+}
+
 void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply& reply)
 {
     if (reply.code == 250) {
         LogPrint(BCLog::TOR, "tor: Authentication successful\n");
 
         // Now that we know Tor is running, possibly set the proxy for onion addresses.
-        bool allow_outbound_onion{true};
-        if (gArgs.IsArgSet("-onlynet")) {
-            allow_outbound_onion = false;
-            for (const auto& net : gArgs.GetArgs("-onlynet")) {
-                if (ParseNetwork(net) == NET_ONION) {
-                    allow_outbound_onion = true;
-                    break;
-                }
-            }
-        }
-        if (gArgs.GetArg("-onion", "") == "" && allow_outbound_onion) {
+        if (gArgs.GetArg("-onion", "").empty() && AllowOutboundOnion()) {
             CService resolved(LookupNumeric("127.0.0.1", 9050));
             proxyType addrOnion = proxyType(resolved, true);
```

Comment (jonatack) [reply src/torcontrol.cpp:387]: Tested with this patch on mainnet:
- `onlynet=i2p`: only I2P outbound connections are made (i2pd 2.23)
- `onlynet=ipv4`: only IPv4 outbound connections are made
- `onlynet=onion`: only onion outbound connections are made (tor 0.4.6)

**Comment by prayank23** [comment]
> See also https://en.wikipedia.org/wiki/Principle_of_least_astonishment

This link says "The behavior should not astonish or surprise users"

I am surprised by lot of things in Bitcoin Core 😄

Concept ACK. Doesn't make sense to create outbound connections with onion peers if `onlynet=i2p` is used.

File: src/torcontrol.cpp (hunk 1 above)
Comment (jonatack) [reply src/torcontrol.cpp:387]: Tested combinations (all tests using the proposed function version ;)
- `onlynet=i2p` and `onlynet=onion`: only I2P and onion outbound connections are made
- `onlynet=ipv4` and `onlynet=onion`: only IPv4 and onion outbound connections are made

(Test coverage for these cases--single onlynet and combinations--would be great.)

File: src/torcontrol.cpp [hunk 2]
```diff
@@ -373,9 +373,18 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
...
             CService resolved(LookupNumeric("127.0.0.1", 9050));
```
Comment (prayank23) [inline src/torcontrol.cpp:388]: This is normally 9151 on Windows so can we use #ifdef here?

```suggestion
#ifdef WIN32 
            CService resolved(LookupNumeric("127.0.0.1", 9150));
#else
            CService resolved(LookupNumeric("127.0.0.1", 9050));
#endif
```

File: src/torcontrol.cpp (hunk 1 above)
Comment (prayank23) [reply src/torcontrol.cpp:387]: > Test coverage for these cases--single onlynet and combinations--would be great

Agree

**Comment by prayank23** [comment]
> It may be good to have functional test coverage (or make it a separate function with unit tests).

Agree. Do you think these steps are correct for test?

1. Run Node 1: 
```
bitcoind -port=18333 -rpcport=18222 -datadir="/home/user/node1" -regtest=1 -listen=1 -server=1 -debug=net -rpcuser=user1 -rpcpassword=password1 -torcontrol='127.0.0.1:9051' -proxy='127.0.0.1:9050' -onlynet=onion
```
2. Run Node 2:
```
bitcoind -port=18777 -rpcport=18666 -datadir="/home/user/node2" -regtest=1 -listen=1 -server=1 -debug=net -rpcuser=user2 -rpcpassword=password2 -addnode='127.0.0.1:18333' -torcontrol='127.0.0.1:9051' -proxy='127.0.0.1:9050' -onlynet=onion
```
3. Stop Node 1 and 2:

```
bitcoin-cli -rpcport=18222 -rpcuser=user1 -rpcpassword=password1 stop
bitcoin-cli -rpcport=18666 -rpcuser=user2 -rpcpassword=password2 stop
```
4. Restart both Node 2 with different config (`onlynet=i2p` and no `proxy`):

```
bitcoind -port=18777 -rpcport=18666 -datadir="/home/user/node2" -regtest=1 -listen=1 -server=1 -debug=net -rpcuser=user2 -rpcpassword=password2 -i2psam=127.0.0.1:7656 -onlynet=i2p
```
5. Check debug.log:

```
trying connection 3llpwlxkisrm2iu5oayh2hd6df7q36wdmt6nounenzqcxkasxfk34eid.onion
```
Master branch should print such connection attempts and PR branch should not

**Comment by jonatack** [comment]
Checking the debug log output is a bit of a last resort when there's nothing else to assert on... getnetworkinfo limited/reachable (and maybe getpeerinfo) would probably be good.

**Comment by wodry** [comment]
Would this fix also #13378 ?

**Comment by prayank23** [comment]
> Checking the debug log output is a bit of a last resort when there's nothing else to assert on... getnetworkinfo limited/reachable (and maybe getpeerinfo) would probably be good.

Also the steps I mentioned above don't work. They worked on one of my VMs. Maybe it already had some onion address in peers.dat. I was expecting it will add onion peer in peers.dat but it didn't. `addpeeraddress` with some onion address and port followed by `getpeerinfo` should work.

**Comment by prayank23** [comment]
Wrote a PowerShell [script ](https://github.com/prayank23/bitcoin-ps-scripts/blob/main/Scripts/Node/onlynet_i2p.ps1) to test this. Tried on Windows 10 and Pop!_OS. Node 2 is connected to Node 1 on Linux and fails on Windows.

TL;DR
   1. Run Node 1 and get onion URL for it. 
   2. Sleep 10 seconds
   3. Get onion address for Node 1 from `getnetworkinfo` -> localaddresses -> address ending with .onion and its port
   4. Run Node 2 with `onlynet=i2p` (No Tor proxy)
   5. Sleep 10 seconds
   6.  Add Node 1 in peers.dat for Node 2 using hidden RPC `addpeeraddress`
   7. Sleep 30 seconds
   8. Check if Node 2 is connected to Node 1 with `getpeerinfo`

```
Peer info:

xejoddwvi35krze3546qtx6dfmednjs2ccsociyezkyhzdxtemp5v5ad.onion:18444
outbound-full-relay
```

File: src/torcontrol.cpp (hunk 2 above)
Comment (vasild) [reply src/torcontrol.cpp:388]: Yes, ouf of scope of this PR. This belongs to https://github.com/bitcoin/bitcoin/pull/15423, I hope that PR moves forward.

**Comment by vasild** [comment]
> Would this fix also #13378 ?

To my understanding, yes, but only if none of `-proxy` or `-onion` is set.

Providing both `-onlynet=ipv4 -onion=127.0.0.1:9050` is kind of contradictory - why provide an outgoing tor proxy if you only want to connect to IPv4? I think Bitcoin Core should print an error and refuse to start in this case, but instead it treats it as "onlynet ipv4 or tor". Changing/fixing that is out of the scope of this PR which only aims to fix the behavior if none of `-proxy` or `-onion` is set.

**Comment by wodry** [comment]
Also of course out of this PR, just to mention: I am sure you are already aware that to have multiple "only" nets is paradox, the option should better be called e.g. "net" instead of "onlynet"..

**Comment by vasild** [comment]
`4f4c7d8b11..4f3020d524`: append a functional test to exercise the modified behavior

File: src/torcontrol.cpp (hunk 1 above)
Comment (vasild) [reply src/torcontrol.cpp:387]: I think it is fine either way. A function would make more sense if this is used in more than one place. The code is now covered by a new functional test `feature_onlynet.py`. Left it as is.

Comment (jonatack) [reply src/torcontrol.cpp:387]: I think the function is not only shorter but also clearer and easier to reason about.

**Comment by vasild** [comment]
`4f3020d524...00b7ba51eb`: hush `test/lint/lint-python.sh`

**Comment by vasild** [comment]
`00b7ba51eb...25f9e113db`: change `allow_outbound_onion` to a function, as per suggestion

File: test/functional/feature_onlynet.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+)
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+)
+
+
+class OnlynetTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.extra_args = [
```
Comment (jonatack) [inline test/functional/feature_onlynet.py:21]: I needed this line for the test to run:
```diff
 class OnlynetTest(BitcoinTestFramework):
     def set_test_params(self):
+        self.setup_clean_chain = True
         self.extra_args = [
```
Otherwise, was seeing the following error:

<details><summary>test output</summary><p>

```
$ test/functional/feature_onlynet.py 
2021-08-11T09:22:40.305000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_p1tdzv8m
2021-08-11T09:22:41.628000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 130, in main
    self.setup()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 285, in setup
    self.setup_network()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 379, in setup_network
    self.setup_nodes()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 409, in setup_nodes
    assert_equal(n.getblockchaininfo()["blocks"], 199)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/util.py", line 49, in assert_equal
    raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
AssertionError: not(0 == 199)
2021-08-11T09:22:41.680000Z TestFramework (INFO): Stopping nodes
2021-08-11T09:22:41.860000Z TestFramework (WARNING): Not cleaning up dir /tmp/bitcoin_func_test_p1tdzv8m
2021-08-11T09:22:41.860000Z TestFramework (ERROR): Test failed. Test logging available at /tmp/bitcoin_func_test_p1tdzv8m/test_framework.log
2021-08-11T09:22:41.861000Z TestFramework (ERROR): 
2021-08-11T09:22:41.861000Z TestFramework (ERROR): Hint: Call /home/jon/projects/bitcoin/bitcoin/test/functional/combine_logs.py '/tmp/bitcoin_func_test_p1tdzv8m' to consolidate all logs
2021-08-11T09:22:41.862000Z TestFramework (ERROR): 
2021-08-11T09:22:41.862000Z TestFramework (ERROR): If this failure happened unexpectedly or intermittently, please file a bug and provide a link or upload of the combined log.
2021-08-11T09:22:41.862000Z TestFramework (ERROR): https://github.com/bitcoin/bitcoin/issues
2021-08-11T09:22:41.862000Z TestFramework (ERROR): 
```
</p></details>

File: test/functional/feature_onlynet.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+        # Passes with or without https://github.com/bitcoin/bitcoin/pull/22651,
+        # just to ensure behavior does not change unintended.
+        n1 = networks_dict(self.nodes[1].getnetworkinfo())
+        assert_equal(n1['onion']['reachable'], True)
+
+        # Passes with or without https://github.com/bitcoin/bitcoin/pull/22651,
+        # just to ensure behavior does not change unintended.
+        n2 = networks_dict(self.nodes[2].getnetworkinfo())
+        assert_equal(n2['onion']['reachable'], True)
```
Comment (jonatack) [inline test/functional/feature_onlynet.py:50]: Here is a proposed simplification using a dict comprehension in `networks_dict` and a loop for the assertions:
```diff
+    def networks_dict(self, node):
+        info = self.nodes[node].getnetworkinfo()
+        return {network["name"]: network for network in info['networks']}
 
     def run_test(self):
-        def networks_dict(d):
-            r = {}
-            for x in d['networks']:
-                r[x['name']] = x
-            return r
-
-        # Would fail without https://github.com/bitcoin/bitcoin/pull/22651
-        n0 = networks_dict(self.nodes[0].getnetworkinfo())
-        assert_equal(n0['onion']['reachable'], False)
-
-        # Passes with or without https://github.com/bitcoin/bitcoin/pull/22651,
-        # just to ensure behavior does not change unintended.
-        n1 = networks_dict(self.nodes[1].getnetworkinfo())
-        assert_equal(n1['onion']['reachable'], True)
-
-        # Passes with or without https://github.com/bitcoin/bitcoin/pull/22651,
-        # just to ensure behavior does not change unintended.
-        n2 = networks_dict(self.nodes[2].getnetworkinfo())
-        assert_equal(n2['onion']['reachable'], True)
+        for node, reachable in [[0, False], [1, True], [2, True]]:
+            assert_equal(self.networks_dict(node)['onion']['reachable'], reachable)
```
Could perhaps add a comment describing why we expect the first node to not be reachable.

File: test/functional/test_framework/basic_server.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        :param bind: Listen address-port tuple, for example ("127.0.0.1", 80)
+
+        :param response_generator: A callback function, supplied with the
+        request. It returns a reply that is to be send to the client either as
+        a string or as an array which is joined into a single string. If it
+        returns None, then the connection is closed by the server.
+        """
+        # create_server() is only in python 3.8
+        # https://docs.python.org/3/library/socket.html#socket.create_server
+        #self.listen_socket = socket.create_server(address=bind, reuse_port=True)
```
Comment (jonatack) [inline test/functional/test_framework/basic_server.py:27]:
```suggestion
        # self.listen_socket = socket.create_server(address=bind, reuse_port=True)
```

File: test/functional/test_framework/basic_server.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        self.listen_socket = socket.socket(socket.AF_INET)
+        self.listen_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        self.listen_socket.bind(bind)
+        self.listen_socket.listen()
+        #
```
Comment (jonatack) [inline test/functional/test_framework/basic_server.py:32]: ?

**Review by jonatack — COMMENTED** [review]
Nice!  A few suggestions for the test.  Only skimmed the basic server so far.

File: test/functional/feature_onlynet.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+# Copyright (c) 2021-2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test -onlynet
+"""
+
+from test_framework.basic_server import (
+    BasicServer,
+    tor_control
```
Comment (jonatack) [inline test/functional/feature_onlynet.py:11]: nit, either add a comma at the end of this line, or one-line it:
```python
from test_framework.basic_server import BasicServer, tor_control
```

**Comment by vasild** [comment]
`test_framework/basic_server.py` can also be used to add some functional tests for I2P. We can simulate an I2P router at least as long as the text/line-based SAM is being talked on the socket. Later, when the socket switches from SAM to the binary bitcoin p2p protocol we would have to close the socket.

File: test/functional/feature_onlynet.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021-2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test -onlynet
```
[the remark this answers is not in the document]
Comment (vasild) [reply test/functional/feature_onlynet.py]: Done.

File: test/functional/feature_onlynet.py (hunk 4 above)
Comment (vasild) [reply test/functional/feature_onlynet.py]: Done something similar. I find `return {network["name"]: network for network in info["networks"]}` extremely hard to read.

File: test/functional/test_framework/basic_server.py (hunk 6 above)
Comment (vasild) [reply test/functional/test_framework/basic_server.py]: I wanted to mark the end of the snippet that is equivalent to `create_server()`. Removed.

File: test/functional/test_framework/basic_server.py (hunk 5 above)
Comment (vasild) [reply test/functional/test_framework/basic_server.py]: Done.

File: test/functional/feature_onlynet.py (hunk 4 above)
Comment (jonatack) [reply test/functional/feature_onlynet.py]: Comprehensions seem idiomatic/pythonic and the most-used way to do it in this codebase too (though I agree that a shorter `n` or `net` for the one line would probably be better than `network`), but I like your new version too. Nice and simple.

File: test/functional/feature_onlynet.py (hunk 3 above)
Comment (jonatack) [reply test/functional/feature_onlynet.py:21]: Hm, after coming back here and rebuilding, the test now runs for me without it. Strange.  Sorry for the noise.

File: test/functional/feature_onlynet.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    def onion_is_reachable(self, node):
+        for net in self.nodes[node].getnetworkinfo()['networks']:
+            if net['name'] == 'onion':
+                return net['reachable']
+        return False
+
+    def run_test(self):
+        # Node 0 would fail the check without https://github.com/bitcoin/bitcoin/pull/22651.
+        # Nodes 1 and 2 pass with or without https://github.com/bitcoin/bitcoin/pull/22651,
+        # they are here just to ensure behavior does not change unintended.
```
Comment (jonatack) [inline test/functional/feature_onlynet.py:41]: nits:

- I think it's more robust to describe why than to refer to a GitHub PR or depend on GitHub in general

- nit s/unintended/unintentionally/

File: src/torcontrol.cpp [hunk 10]
```diff
@@ -368,14 +368,26 @@ void TorController::add_onion_cb(TorControlConnection& _conn, const TorControlRe
...
 void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply& reply)
 {
     if (reply.code == 250) {
         LogPrint(BCLog::TOR, "tor: Authentication successful\n");
 
-        // Now that we know Tor is running setup the proxy for onion addresses
-        // if -onion isn't set to something else.
-        if (gArgs.GetArg("-onion", "") == "") {
+        // Now that we know Tor is running, possibly set the proxy for onion addresses.
+        if (gArgs.GetArg("-onion", "") == "" && AllowOutboundOnion()) {
```
Comment (jonatack) [inline src/torcontrol.cpp:390]:
```suggestion
        if (gArgs.GetArg("-onion", "").empty() && AllowOutboundOnion()) {
```

File: test/functional/feature_onlynet.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+
+class OnlynetTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.extra_args = [
+            ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1'],
+            ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-onion=127.0.0.1:9050'],
+            ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-proxy=127.0.0.1:9050'],
+        ]
```
Comment (jonatack) [inline test/functional/feature_onlynet.py:25]: Would it make sense to test with multiple onlynet settings?

```diff
diff --git a/test/functional/feature_onlynet.py b/test/functional/feature_onlynet.py
index 60e4caf33d..c43a541196 100755
--- a/test/functional/feature_onlynet.py
+++ b/test/functional/feature_onlynet.py
@@ -22,6 +22,8 @@ class OnlynetTest(BitcoinTestFramework):
             ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1'],
             ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-onion=127.0.0.1:9050'],
             ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-proxy=127.0.0.1:9050'],
+            ['-onlynet=ipv4', '-onlynet=ipv6', '-torcontrol=127.0.0.1:19051', '-listenonion=1'],
         ]
         self.num_nodes = len(self.extra_args)

@@ -39,7 +41,7 @@ class OnlynetTest(BitcoinTestFramework):
-        for node, expected_reachable in [[0, False], [1, True], [2, True]]:
+        for node, expected_reachable in [[0, False], [1, True], [2, True], [3, False]]:
             self.log.info(f'Verifying node {node} which uses {self.extra_args[node]}')
             assert_equal(self.onion_is_reachable(node), expected_reachable)
```

**Review by jonatack — COMMENTED** [review]
ACK 16bf93e1361fa9a1503cadd96e5121d92937bfb5 rebased to master, debug build, ran new tests, verified that the test fails where expected on master, and previously manually verified the behavior extensively as described in https://github.com/bitcoin/bitcoin/pull/22651#discussion_r684429929 and https://github.com/bitcoin/bitcoin/pull/22651#discussion_r684434758

Happy to re-ACK for further improvements.

**Comment by Talkless** [comment]
Concept ACK

**Review by ghost — APPROVED** [review]
tACK https://github.com/bitcoin/bitcoin/pull/22651/commits/16bf93e1361fa9a1503cadd96e5121d92937bfb5

Tested with [test/functional/feature_onlynet.py](https://raw.githubusercontent.com/bitcoin/bitcoin/16bf93e1361fa9a1503cadd96e5121d92937bfb5/test/functional/feature_onlynet.py) and my [PowerShell Script](https://github.com/bitcoin/bitcoin/pull/22651#issuecomment-894761007)

Master Branch:

1. Python test fails
```
2021-08-11T22:47:24.271000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_tfz0x_fp
2021-08-11T22:47:26.786000Z TestFramework (INFO): Verifying node 0 which uses ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1']
2021-08-11T22:47:26.790000Z TestFramework (ERROR): Assertion failed
```
2. PS script returns : `xejoddwvi35krze3546qtx6dfmednjs2ccsociyezkyhzdxtemp5v5ad.onion:18444 outbound-full-relay`

PR Branch:

1. Python test successful:
```
11T22:44:25.360000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_6vo22aqz
2021-08-11T22:44:26.622000Z TestFramework (INFO): Verifying node 0 which uses ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1']
2021-08-11T22:44:26.623000Z TestFramework (INFO): Verifying node 1 which uses ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-onion=127.0.0.1:9050']
2021-08-11T22:44:26.625000Z TestFramework (INFO): Verifying node 2 which uses ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-proxy=127.0.0.1:9050']
2021-08-11T22:44:26.629000Z TestFramework (INFO): Stopping nodes
2021-08-11T22:44:26.791000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_6vo22aqz on exit
2021-08-11T22:44:26.791000Z TestFramework (INFO): Tests successful
    
```
2. PS script returns nothing (no onion peer)

Although both tests use different approach and a basic TCP server is added in test framework for this test, most important commit is https://github.com/bitcoin/bitcoin/pull/22651/commits/c591379eb03381145c85ee5760ce829be337a749 which looks good and fixes the issue.

Thanks for fixing this issue and improving privacy in Bitcoin Core. Hope this is merged soon and tests can always be improved in future if required.

**Comment by vasild** [comment]
`16bf93e136...5959ece29e`: address minor suggestions and avoid possibility of colliding listening ports by binding to port 0 (let the OS choose an available port, thanks @darosior for the idea!)

File: test/functional/feature_onlynet.py (hunk 11 above)
Comment (vasild) [reply test/functional/feature_onlynet.py:28]: I think it is fine as it is. The point is to have `onlynet=...` but not `onlynet=onion`.

File: src/torcontrol.cpp (hunk 10 above)
Comment (vasild) [reply src/torcontrol.cpp:390]: You suggested this one earlier and I deliberately skipped it but did not explain why.

Here we are testing whether the result of `GetArg()` equals to its second argument, so it is better to keep that obvious, e.g. `GetArg(..., X) == X`.

File: test/functional/feature_onlynet.py (hunk 9 above)
Comment (vasild) [reply test/functional/feature_onlynet.py]: I agree, removed the github reference.

I think "uninternationally" is even better, changed.

File: test/functional/feature_onlynet.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    def onion_is_reachable(self, node):
+        for net in self.nodes[node].getnetworkinfo()['networks']:
+            if net['name'] == 'onion':
+                return net['reachable']
+        return False
+
+    def run_test(self):
+        # Node 0 would fail the check without c591379eb03381145c85ee5760ce829be337a749.
+        # Nodes 1 and 2 pass with or without that commit, they are here just to ensure
+        # behavior does not change unintentionally.
```
Comment (jonatack) [inline test/functional/feature_onlynet.py:40]: maybe use a docstring and describe the commit, feel free to ignore
```diff
-        # Node 0 would fail the check without c591379eb03381145c85ee5760ce829be337a749.
-        # Nodes 1 and 2 pass with or without that commit, they are here just to ensure
-        # behavior does not change unintentionally.
+        """
+        - The test of node 0 would fail without commit c591379eb0,
+          "tor: respect non-onion -onlynet= for outgoing Tor connections."
+
+        - The tests of nodes 1 and 2 pass with or without that commit;
+          they ensure that behavior does not change unintentionally.
+        """
```

File: test/functional/feature_onlynet.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        for node, expected_reachable in [[0, False], [1, True], [2, True]]:
+            self.log.info(f'Verifying node {node} which uses {self.extra_args[node]}')
```
Comment (jonatack) [inline test/functional/feature_onlynet.py:42]: output is fairly long, could shorten without loss of signal
```suggestion
            self.log.info(f'Test node {node} {self.extra_args[node]}')
```

File: test/functional/test_framework/basic_server.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+import socket
+import threading
+
+
+class BasicServer:
+    def __init__(self, *, bind, response_generator):
+        """
+        :param bind: Listen address-port tuple, for example ("127.0.0.1", 80).
+        If the port is 0 then a random available one will be chosen and it
```
Comment (jonatack) [inline test/functional/test_framework/basic_server.py:17]: indeed, seeing a different random ports as described here

**Review by jonatack — COMMENTED** [review]
ACK 5959ece29e5b6b7ce0d8be60ae64a3312861f841 tested rebased to master, verified that the test fails where expected on master, and previously manually verified the behavior extensively as described in https://github.com/bitcoin/bitcoin/pull/22651#discussion_r684429929 and https://github.com/bitcoin/bitcoin/pull/22651#discussion_r684434758

File: src/torcontrol.cpp (hunk 10 above)
Comment (Talkless) [reply src/torcontrol.cpp:390]: Because `==` will generate/call string comparison code, instead of just checking string length using `empty()`. Better just use specialized member functions instead of some kind "generic" comparison.

Another example is using `clear()` vs `= ""`: https://www.youtube.com/watch?v=3X9qK7HWxjk

File: src/torcontrol.cpp [hunk 15]
```diff
@@ -368,14 +368,26 @@ void TorController::add_onion_cb(TorControlConnection& _conn, const TorControlRe
     }
 }
 
+static bool AllowOutboundOnion()
+{
+    if (!gArgs.IsArgSet("-onlynet")) {
+        return true;
+    }
+    for (const auto& net : gArgs.GetArgs("-onlynet")) {
```
Comment (Talkless) [inline src/torcontrol.cpp:376]: Please consider using standard algorithms instead of hand-rolled loop:
```
#include <algorithm>
...

    const auto only_nets{gArgs.GetArgs("-onlynet")};

    //TODO: C++20: use std::ranges::any_of() instead
    return std::any_of(only_nets.cbegin(), only_nets.cend(), [](const auto &net) {
        return ParseNetwork(net) == NET_ONION;
    });
```

**Comment by jonatack** [comment]
This has a couple of ACKs and would be very good to have in 22.0 to alleviate user confusion over an impactful user-facing change in the release. @vasild, do you plan to update? Anyone else like to review here?

File: src/torcontrol.cpp (hunk 10 above)
Comment (vasild) [reply src/torcontrol.cpp:390]: I think readability trumps micro-optimization here, leaving it as is.

File: test/functional/feature_onlynet.py (hunk 12 above)
Comment (vasild) [reply test/functional/feature_onlynet.py:40]: Leaving it as is.

File: test/functional/feature_onlynet.py (hunk 13 above)
Comment (vasild) [reply test/functional/feature_onlynet.py:42]: Done.

File: src/torcontrol.cpp (hunk 15 above)
Comment (vasild) [reply src/torcontrol.cpp:376]: Done.

**Comment by vasild** [comment]
`5959ece29e...a2ebcda19a`: address some suggestions

> @vasild, do you plan to update?

Done! :)

**Comment by jonatack** [comment]
Code review re-ACK a2ebcda19a25067cb8183f25f6ffd0a26065fcc5 per `git diff 5959ece a2ebcda` plus sanity check of rebase on master, debug build, re-ran test a few times

**Comment by Rspigler** [comment]
Concept ACK

**Comment by prayank23** [comment]
reACK https://github.com/bitcoin/bitcoin/pull/22651/commits/a2ebcda19a25067cb8183f25f6ffd0a26065fcc5

Major changes since last review:

1. https://github.com/bitcoin/bitcoin/pull/22651#discussion_r689088453
2. https://github.com/bitcoin/bitcoin/pull/22651#issuecomment-897498176

File: src/torcontrol.cpp [hunk 16]
```diff
@@ -368,14 +369,24 @@ void TorController::add_onion_cb(TorControlConnection& _conn, const TorControlRe
     }
 }
 
+static bool AllowOutboundOnion()
+{
+    if (!gArgs.IsArgSet("-onlynet")) {
+        return true;
+    }
+    const auto& onlynets = gArgs.GetArgs("-onlynet");
```
[the remark this answers is not in the document]
Comment (vasild) [reply src/torcontrol.cpp:377]: Removed.

**Comment by vasild** [comment]
`a2ebcda19a...baa6cb12ac`: address minor suggestion and fix commit id in a comment.

**Comment by vasild** [comment]
@laanwj, I agree with your comments at https://github.com/bitcoin/bitcoin/pull/22648#discussion_r696535966. Opened https://github.com/bitcoin/bitcoin/pull/22834 which if merged will make this PR unnecessary.

**Comment by jonatack** [comment]
Code review re-ACK baa6cb12ac5071bb68eb5ec375be73df4f4ae28f per `git diff a2ebcda baa6cb1`, rebase on master, debug build, ran the new test `test/functional/feature_onlynet.py` a few times, previously manually verified the behavior extensively as described in https://github.com/bitcoin/bitcoin/pull/22651#discussion_r684429929 and https://github.com/bitcoin/bitcoin/pull/22651#discussion_r684434758

Will review #22834 soon. Perhaps the functional test work here can be ~ported over to it if that change is preferred.

**Comment by luke-jr** [comment]
#22834 looks like a better solution IMO

**Comment by Talkless** [comment]
> #22834 looks like a better solution IMO

So this PR should be closed?

@vasild I'm confused :)

**Comment by vasild** [comment]
Closing this in favor of https://github.com/bitcoin/bitcoin/pull/22834 which has more (concept) ACKs.

Thanks, @Talkless!
